File: java/1352.java
class ProductOfNumbers {
    static final int pa[] = new int[40001];
    int i, l = -1;

    public ProductOfNumbers() {
        pa[0] = 1;
        i = 0;
    }

    public void add(int num) {
        i++;
        if (num == 0) {
            l = i;
            pa[i] = 1;
        } else
            pa[i] = pa[i - 1] * num;
    }

    public int getProduct(int k) {
        k = i - k;
        if (l > k)
            return 0;
        return pa[i] / pa[k];
    }
}

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers obj = new ProductOfNumbers();
 * obj.add(num);
 * int param_2 = obj.getProduct(k);
 */

File: python3/1352.py
class ProductOfNumbers:

    def __init__(self):
        self.pa = [1]
        self.i = 0

    def add(self, num: int) -> None:
        if num == 0:
            self.pa = [1]
            self.i = 0
        else:
            self.pa.append(self.pa[self.i] * num)
            self.i += 1

    def getProduct(self, k: int) -> int:
        if k > self.i:
            return 0
        return int(self.pa[-1] / self.pa[-k - 1])


# Your ProductOfNumbers object will be instantiated and called as such:
# obj = ProductOfNumbers()
# obj.add(num)
# param_2 = obj.getProduct(k)
